### PR TITLE
Make federation catchup send last event from any server.

### PR DESCRIPTION
**Pull request by erikjohnston**
Currently federation catchup will send the last *local* event that we failed to send to the remote. This can cause issues for large rooms where lots of servers have sent events while the remote server was down, as when it comes back up again it'll be flooded with events from various points in the DAG.

Instead, let's make it so that all the servers send the most recent events, even if its not theirs. The remote should deduplicate the
events, so there shouldn't be much overhead in doing this. Alternatively, the servers could only send local events if they were also extremities and hope that the other server will send the event over, but that is a bit risky.

c.f. https://github.com/matrix-org/synapse/issues/9490 and https://github.com/matrix-org/synapse/pull/9621

### CODE REVIEW

File: synapse/federation/sender/per_destination_queue.py [hunk 1]
```diff
@@ -415,22 +416,93 @@ async def _catch_up_transmission_loop(self) -> None:
...
-                rooms = [p.room_id for p in catchup_pdus]
-                logger.info("Catching up rooms to %s: %r", self._destination, rooms)
+            # We send transactions with events from one room only, as its likely
+            # that the remote will have to do additional processing, which may
+            # take some time. It's better to give it small amounts of work
+            # rather than risk the request timing out and repeatedly being
+            # retried, and not making any progress.
+            for pdu in catchup_pdus:
+                # The PDU from the DB will be the last PDU in the room from
+                # *this server* that wasn't sent to the remote. However, other
```
Comment (clokep) [inline synapse/federation/sender/per_destination_queue.py:426]: This is because we only know that *we* failed to send an event, but the remote server might be communicating OK with other servers in the room? (Meaning that it could actually already have the event we're trying to send to it?)

File: synapse/federation/sender/per_destination_queue.py [hunk 2]
```diff
@@ -415,22 +416,93 @@ async def _catch_up_transmission_loop(self) -> None:
...
 
-            if logger.isEnabledFor(logging.INFO):
-                rooms = [p.room_id for p in catchup_pdus]
-                logger.info("Catching up rooms to %s: %r", self._destination, rooms)
+            # We send transactions with events from one room only, as its likely
+            # that the remote will have to do additional processing, which may
+            # take some time. It's better to give it small amounts of work
+            # rather than risk the request timing out and repeatedly being
+            # retried, and not making any progress.
+            for pdu in catchup_pdus:
```
Comment (clokep) [inline synapse/federation/sender/per_destination_queue.py:424]: Should we be grouping these by room first? Otherwise it seems like we would do some unnecessary work. (I think at least...)

File: synapse/federation/federation_server.py [hunk 3]
```diff
@@ -727,27 +727,6 @@ async def _handle_received_pdu(self, origin: str, pdu: EventBase) -> None:
             if the event was unacceptable for any other reason (eg, too large,
             too many prev_events, couldn't find the prev_events)
         """
-        # check that it's actually being sent from a valid destination to
-        # workaround bug #1753 in 0.18.5 and 0.18.6
-        if origin != get_domain_from_id(pdu.sender):
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/federation/federation_server.py:732]: This needs to get removed as we're now sending events in transactions that aren't from the originating server. It's safe to remove due to the bug having long since been fixed.

File: synapse/federation/sender/per_destination_queue.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/federation/sender/per_destination_queue.py:426]: Err, I don't follow really. The server will only track the last event it failed to send to the remote was, which will only be local events. When the server comes back up we want it to be told about the most recent events in the room, so that it can figure out it missed stuff. We don't want to send anything other than the latest events as that will cause unnecessary processing on the remote, and the last event sent by the server to a room won't necessarily be the part of the latest events in the room (as other servers may have sent events since).

In other words, we want the remote to only receive the current forward extremities in the room; we're not trying to send it all events its missed.

Comment (clokep) [reply synapse/federation/sender/per_destination_queue.py:426]: > The server will only track the last event it failed to send to the remote was, which will only be local events.

This is what I was trying to confirm. 👍

**Review by clokep — APPROVED** [review]
